### PR TITLE
Implement adoption of k8s clusters

### PR DESCRIPTION
**Pull request by kylewuolle**
This PR resolves issue https://github.com/Mirantis/hmc/issues/542 by:

1. Renaming ManagedCluster to ClusterDeployment
2. Creation of a new template called adopted cluster that allows creating services via a sveltoscluster for existing k8s clusters not created via 2a

I've added a PR to update the documentation : https://github.com/Mirantis/project-2a-docs/pull/68

### CODE REVIEW

**Comment by zerospiel** [comment]
@kylewuolle can the PR be split into two different PRs (or at least 2 different commits within the PR): the one with the renaming and the one with the adopted templates feature? Otherwise, there are too many changes to review and each requires attention because one might be either a simple renaming (which is pretty simple to review) or a feature (which requires extra attention)

**Comment by kylewuolle** [comment]
> @kylewuolle can the PR be split into two different PRs (or at least 2 different commits within the PR): the one with the renaming and the one with the adopted templates feature? Otherwise, there are too many changes to review and each requires attention because one might be either a simple renaming (which is pretty simple to review) or a feature (which requires extra attention)

Done!

**Review by zerospiel — COMMENTED** [review]
Mostly, minor comments. I have a generic question: so the whole approach is to provide a template with an external (third-party, BYO) cluster, so it could be attached to the HMC but without handling any actions related to the provisioning and its managing, right? 

I also have a concern regarding the provider name `infrastracture-internal`, I am not aware how it should be named in terms of CAPI, I'd consult with other team members regarding it

File: templates/provider/hmc/templates/crds/hmc.mirantis.com_managedclusters.yaml [hunk 1]
```diff
@@ -1,295 +0,0 @@
----
```
[the remark this answers is not in the document]
Comment (zerospiel) [reply templates/provider/hmc/templates/crds/hmc.mirantis.com_managedclusters.yaml]: outdated, it was related to the first commit, so I close it

File: api/v1alpha1/managedcluster_types.go [hunk 2]
```diff
@@ -111,27 +111,27 @@ type ManagedClusterStatus struct {
 // +kubebuilder:printcolumn:name="status",type="string",JSONPath=".status.conditions[?(@.type==\"Ready\")].message",description="Status",priority=0
 // +kubebuilder:printcolumn:name="dryRun",type="string",JSONPath=".spec.dryRun",description="Dry Run",priority=1
 
-// ManagedCluster is the Schema for the managedclusters API
-type ManagedCluster struct {
+// ClusterDeployment is the Schema for the ClusterDeployments API
```
[the remark this answers is not in the document]
Comment (kylewuolle) [reply api/v1alpha1/managedcluster_types.go:114]: Done!

File: internal/telemetry/event.go [hunk 3]
```diff
@@ -21,30 +21,30 @@ import (
 )
 
 const (
-	managedClusterCreateEvent    = "managed-cluster-create"
-	managedClusterHeartbeatEvent = "managed-cluster-heartbeat"
+	ClusterDeploymentCreateEvent    = "cluster-deployment-create"
+	ClusterDeploymentHeartbeatEvent = "cluster-deployment-heartbeat"
```
[the remark this answers is not in the document]
Comment (kylewuolle) [reply internal/telemetry/event.go]: Done!

File: api/v1alpha1/indexers.go [hunk 4]
```diff
@@ -85,17 +85,17 @@ func ExtractServiceTemplateNamesFromManagedCluster(rawObj client.Object) []strin
...
+	return mgr.GetFieldIndexer().IndexField(ctx, &ClusterDeployment{}, ClusterDeploymentCredentialIndexKey, ExtractCredentialNameFromClusterDeployment)
 }
 
-// ExtractCredentialNameFromManagedCluster returns referenced Credential name
-// declared in a ManagedCluster object.
-func ExtractCredentialNameFromManagedCluster(rawObj client.Object) []string {
-	cluster, ok := rawObj.(*ManagedCluster)
+// ExtractCredentialNameFromClusterDeployment returns referenced Credential name
+// declared in a ClusterDeployment object.
+func ExtractCredentialNameFromClusterDeployment(rawObj client.Object) []string {
```
[the remark this answers is not in the document]
Comment (kylewuolle) [reply api/v1alpha1/indexers.go]: Done!

File: test/objects/clusterdeployment/clusterdeployment.go [hunk 5]
```diff
@@ -22,14 +22,14 @@ import (
 )
 
 const (
-	DefaultName      = "managedcluster"
+	DefaultName      = "clusterdeployment"
 	DefaultNamespace = metav1.NamespaceDefault
 )
 
-type Opt func(managedCluster *v1alpha1.ManagedCluster)
+type Opt func(ClusterDeployment *v1alpha1.ClusterDeployment)
```
[the remark this answers is not in the document]
Comment (kylewuolle) [reply test/objects/clusterdeployment/clusterdeployment.go]: Done!

File: Makefile [hunk 6]
```diff
@@ -334,6 +334,10 @@ dev-templates: templates-generate
 dev-release:
 	@$(YQ) e ".spec.version = \"${VERSION}\"" $(PROVIDER_TEMPLATES_DIR)/hmc-templates/files/release.yaml | $(KUBECTL) -n $(NAMESPACE) apply -f -
 
+.PHONY: dev-adopted-creds
```
[the remark this answers is not in the document]
Comment (kylewuolle) [reply Makefile:337]: I've got it in the dev docs

File: config/dev/adopted-clusterdeployment.yaml [hunk 7]
```diff
@@ -0,0 +1,16 @@
+apiVersion: hmc.mirantis.com/v1alpha1
+kind: ClusterDeployment
+metadata:
+  name: aws-dev
```
[the remark this answers is not in the document]
Comment (kylewuolle) [reply config/dev/adopted-clusterdeployment.yaml]: done!

File: internal/controller/management_controller.go [hunk 8]
```diff
@@ -123,6 +123,9 @@ func (r *ManagementReconciler) Update(ctx context.Context, management *hmc.Manag
 
 		requeue bool
 	)
+
+	statusAccumulator.providers = append(statusAccumulator.providers, "infrastructure-internal")
```
[the remark this answers is not in the document]
Comment (kylewuolle) [reply internal/controller/management_controller.go]: done!

File: templates/provider/hmc/templates/deployment.yaml [hunk 9]
```diff
@@ -54,13 +54,15 @@ spec:
             port: 8081
           initialDelaySeconds: 15
           periodSeconds: 20
+          timeoutSeconds: 6000
```
[the remark this answers is not in the document]
Comment (kylewuolle) [reply templates/provider/hmc/templates/deployment.yaml]: Sorry that was some test stuff I left in by mistake thanks for catching!

File: internal/controller/clusterdeployment_controller.go [hunk 10]
```diff
@@ -69,46 +69,48 @@ type ManagedClusterReconciler struct {
...
 	}
 
-	return r.reconcileUpdate(ctx, managedCluster)
+	return r.reconcileUpdate(ctx, clusterDeployment)
 }
 
-func (r *ManagedClusterReconciler) setStatusFromChildObjects(ctx context.Context, managedCluster *hmc.ManagedCluster, gvr schema.GroupVersionResource, conditions []string) (requeue bool, _ error) {
+func (r *ClusterDeploymentReconciler) setStatusFromChildObjects(
+	ctx context.Context, clusterDeployment *hmc.ClusterDeployment, gvr schema.GroupVersionResource, conditions []string,
+) (requeue bool, _ error) {
```
[the remark this answers is not in the document]
Comment (kylewuolle) [reply internal/controller/clusterdeployment_controller.go]: Old habits from the MSR team where long lines were not allowed

**Review by zerospiel — APPROVED** [review]
LGTM, I'd wait for the ideas regarding the `infrastructure-internal` provider name and maybe update the `ClusterDeployment` aliases to something like `cldeploy` (the only alias, `cld` might potentially interfere with other resources one day) but it is a personal preference (`clusterd` sounds similar to a cluster daemon)

**Review by eromanova — COMMENTED** [review]
Mostly minor comments.
Ideally, before creating the `ClusterDeployment` with the adopted template, we should ensure that Sveltos is ready (because the template includes objects from the Sveltos API). For example, consider the template as invalid if the Sveltos is not ready yet (by the template controller) and then validate it in the webhook on a `clusterdeployment` creation, similarly to the current approach with the providers' readiness. But sveltos is not a provider and we probably should have a separate annotation for that and implement a separate logic in the controller. Anyway, It’s worth gathering more opinions on this approach.

**Review by zerospiel — APPROVED** [review]

**Comment by mestadler** [comment]
@Kshatrix Agreed, this looks good from my perspective, please go ahead.

**Review by zerospiel — APPROVED** [review]

**Pull request merged**